File: bezier/src/org/brogan/data/OvalSetXml.java
package org.brogan.data;

import java.io.File;
import java.io.FileOutputStream;
import java.util.List;

import org.brogan.bezier.OvalManager;

import nu.xom.Attribute;
import nu.xom.Document;
import nu.xom.Element;
import nu.xom.Serializer;

/**
 * Saves a list of OvalManagers as an <ovalSet> XML file.
 * Coordinates are normalised to the -0.5..+0.5 space used by all other geometry types:
 *   normalised = (canvas_px - edgeOffset) / gridSize - 0.5
 * where gridSize = 1000 and edgeOffset = 20.
 *
 * File is written without a DOCTYPE — loaded with a non-validating parser.
 */
public class OvalSetXml {

    private static final double GRID = 1000.0;
    private static final double EDGE = 20.0;

    private String xmlFilePath;

    public void setXmlFilePath(String path) { this.xmlFilePath = path; }

    /**
     * Create and save the XML file.
     * @param name       base name (used in <name> element)
     * @param ovals      list of OvalManagers to serialise
     */
    public void createNewXml(String name, List<OvalManager> ovals) {
        Element root = new Element("ovalSet");

        Element nameEl = new Element("name");
        nameEl.appendChild(name);
        root.appendChild(nameEl);

        for (OvalManager oval : ovals) {
            Element ovalEl = new Element("oval");
            ovalEl.addAttribute(new Attribute("cx", fmt(toNorm(oval.getCx()))));
            ovalEl.addAttribute(new Attribute("cy", fmt(toNorm(oval.getCy()))));
            ovalEl.addAttribute(new Attribute("rx", fmt(toNormRadius(oval.getRx()))));
            ovalEl.addAttribute(new Attribute("ry", fmt(toNormRadius(oval.getRy()))));
            root.appendChild(ovalEl);
        }

        Document doc = new Document(root);
        try {
            new File(xmlFilePath).getParentFile().mkdirs();
            FileOutputStream fos = new FileOutputStream(new File(xmlFilePath));
            Serializer out = new Serializer(fos, "UTF-8");
            out.setIndent(4);
            out.write(doc);
            fos.close();
        } catch (Exception e) {
            System.err.println("OvalSetXml: failed to save: " + e.getMessage());
        }
    }

    // ── Coordinate helpers ────────────────────────────────────────────────

    /** Canvas pixel → normalised centre coordinate (-0.5..+0.5). */
    private static double toNorm(double px) {
        return (px - EDGE) / GRID - 0.5;
    }

    /** Canvas pixel radius → normalised radius (0..0.5). */
    private static double toNormRadius(double r) {
        return r / GRID;
    }

    private static String fmt(double v) {
        // Round to 4 decimal places
        return String.format("%.4f", v);
    }
}

File: bezier_py/bezier_io/oval_set_xml.py
"""
OvalSet XML IO — read/write ovalSet XML files.
No DOCTYPE. UTF-8 encoding.

Coordinate pipeline (matches Java OvalSetXml):
  Save centre: norm = (canvas_px - EDGE) / GRID - 0.5
  Save radius:  norm = radius / GRID
  Load centre:  canvas_px = (norm + 0.5) * GRID + EDGE
  Load radius:  radius = norm * GRID
"""
from __future__ import annotations
import os
from xml.etree import ElementTree as ET
from model.oval_manager import OvalManager

GRID = 1000.0
EDGE = 20.0
XML_DECL = '<?xml version="1.0" encoding="UTF-8"?>'


def _to_norm(canvas_px: float) -> str:
    v = (canvas_px - EDGE) / GRID - 0.5
    return f"{v:.4f}"

def _to_norm_radius(r: float) -> str:
    return f"{r / GRID:.4f}"

def _from_norm(norm: float) -> float:
    return (norm + 0.5) * GRID + EDGE

def _from_norm_radius(norm: float) -> float:
    return norm * GRID


def write_oval_set(file_path: str, name: str, ovals: list[OvalManager]) -> None:
    """Write an ovalSet XML file (no DOCTYPE, UTF-8)."""
    lines: list[str] = [
        XML_DECL,
        '<ovalSet>',
        f'    <name>{_esc(name)}</name>',
    ]
    for oval in ovals:
        cx = _to_norm(oval.cx)
        cy = _to_norm(oval.cy)
        rx = _to_norm_radius(oval.rx)
        ry = _to_norm_radius(oval.ry)
        lines.append(f'    <oval cx="{cx}" cy="{cy}" rx="{rx}" ry="{ry}"/>')
    lines.append('</ovalSet>')

    os.makedirs(os.path.dirname(os.path.abspath(file_path)), exist_ok=True)
    with open(file_path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(lines) + '\n')


def read_oval_set(file_path: str) -> list[OvalManager]:
    """Parse an ovalSet XML and return a list of OvalManagers in canvas pixel space."""
    with open(file_path, 'r', encoding='utf-8', errors='replace') as f:
        content = f.read()
    import re
    content = re.sub(r'<!DOCTYPE[^>]*>', '', content)
    root = ET.fromstring(content)
    ovals: list[OvalManager] = []
    for el in root.findall('oval'):
        cx = _from_norm(float(el.get('cx', '0')))
        cy = _from_norm(float(el.get('cy', '0')))
        rx = _from_norm_radius(float(el.get('rx', '0')))
        ry = _from_norm_radius(float(el.get('ry', '0')))
        ovals.append(OvalManager(cx, cy, rx, ry))
    return ovals


def _esc(s: str) -> str:
    return s.replace('&', '&amp;').replace('<', '&lt;').replace('>', '&gt;')
